runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -135,31 +149,131 @@ public void open() throws Exception {
 
+    final StateRequestHandler sideInputStateHandler;
     if (executableStage.getSideInputs().size() > 0) {
       checkNotNull(super.sideInputHandler);
       StateRequestHandlers.SideInputHandlerFactory sideInputHandlerFactory =
           Preconditions.checkNotNull(
               FlinkStreamingSideInputHandlerFactory.forStage(
                   executableStage, sideInputIds, super.sideInputHandler));
       try {
-        return StateRequestHandlers.forSideInputHandlerFactory(
-            ProcessBundleDescriptors.getSideInputs(executableStage), sideInputHandlerFactory);
+        sideInputStateHandler =
+            StateRequestHandlers.forSideInputHandlerFactory(
+                ProcessBundleDescriptors.getSideInputs(executableStage), sideInputHandlerFactory);
       } catch (IOException e) {
-        throw new RuntimeException(e);
+        throw new RuntimeException("Failed to initialize SideInputHandler", e);
+      }
+    } else {
+      sideInputStateHandler = StateRequestHandler.unsupported();
+    }
+
+    final StateRequestHandler userStateRequestHandler;
+    if (executableStage.getUserStates().size() > 0) {
+      if (keyedStateInternals == null) {
+        throw new IllegalStateException("Input must be keyed when user state is used");
       }
+      userStateRequestHandler =
+          StateRequestHandlers.forBagUserStateHandlerFactory(
+              stageBundleFactory.getProcessBundleDescriptor(),
+              new BagUserStateFactory(keyedStateInternals, getKeyedStateBackend()));
     } else {
-      return StateRequestHandler.unsupported();
+      userStateRequestHandler = StateRequestHandler.unsupported();
+    }
+
+    EnumMap<TypeCase, StateRequestHandler> handlerMap = new EnumMap<>(TypeCase.class);
+    handlerMap.put(TypeCase.MULTIMAP_SIDE_INPUT, sideInputStateHandler);
+    handlerMap.put(TypeCase.BAG_USER_STATE, userStateRequestHandler);
+
+    return StateRequestHandlers.delegateBasedUponType(handlerMap);
+  }
+
+  private static class BagUserStateFactory
+      implements StateRequestHandlers.BagUserStateHandlerFactory {
+
+    private final StateInternals stateInternals;
+    private final KeyedStateBackend<ByteBuffer> keyedStateBackend;
+
+    private BagUserStateFactory(
+        StateInternals stateInternals, KeyedStateBackend<ByteBuffer> keyedStateBackend) {
+
+      this.stateInternals = stateInternals;
+      this.keyedStateBackend = keyedStateBackend;
+    }
+
+    @Override
+    public <K, V, W extends BoundedWindow>
+        StateRequestHandlers.BagUserStateHandler<K, V, W> forUserState(
+            String pTransformId,
+            String userStateId,
+            Coder<K> keyCoder,
+            Coder<V> valueCoder,
+            Coder<W> windowCoder) {
+      return new StateRequestHandlers.BagUserStateHandler<K, V, W>() {
+        @Override
+        public Iterable<V> get(K key, W window) {
+          prepareStateBackend(key, keyCoder);
```

Comment:
Is it possible for multiple threads to serve state requests concurrently? Do we need to synchronize access to the state request handler?

Reply:
Good question. I thought the GrpcStateService is an actor which processes incoming state requests single-threaded. Since state requests itself are synchronously processed we should have be fine then. 

However, GrpcStateService can handle requests from multiple clients concurrently, but there is not more than one connection for each Harness. So when multiple harnesses are connected, there may be concurrent requests, which would not be a problem because they access different operator state.

ExecutableStages should not be chained together because that would break the assumption that they run with independent operator state. So in the case where two chained ExecutableStages each used their own SDK harness it would be a problem. AFAIK that shouldn't happen. Perhaps we should disable chaining for stateful execution, just to be sure.

Reply:
Just realized that I commented on the commit instead of continuing the thread:

There are legitimate cases for chaining with different operator upstream of the executable stage (such as a native transform like the Kinesis consumer)

It is only a concern for executable stages that use stateful processing. If KeyedStateBackend is shared between operators in a chain, then we should set the chaining strategy only in such case.

It is currently possible that executable stages (chained or not) are served in different sdk workers and therefore the scenario you describe is possible. An alternative to trying to work out when exactly to allow chaining or not would be to just synchronize access to the state backend.

Reply:
I've looked into the chaining code and apparently the KeyedStataBackend is scoped by operator id. So it shouldn't conflict, even if the operators are chained together.

I disabled chaining because I have doubt that the asynchronous processing of multiple elements (now that we merged #6723) works with other chained operators, because it might break the chain abstraction. I'll remove the chaining change and will investigate that separately.